File: packages/contracts/src/catalyst_contracts/result.py
"""Backtest result contract models (backtest-result.schema.json)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import Field

from ._base import Decimal, StrictModel
from .policy import SimulationPolicy
from .trace import Portfolio


class Summary(StrictModel):
    starting_value_usd: Decimal
    final_value_usd: Decimal
    pnl_usd: Decimal
    return_pct: Decimal
    max_drawdown_pct: Decimal | None = None
    trade_count: int | None = None
    rejected_count: int | None = None


class EquityPoint(StrictModel):
    ts: datetime
    equity_usd: Decimal


class DrawdownPoint(StrictModel):
    ts: datetime
    drawdown_pct: Decimal


class Trade(StrictModel):
    ts: datetime
    node_id: str
    kind: str
    venue: str | None = None
    symbol: str | None = None
    side: str | None = None
    price: Decimal | None = None
    amount: Decimal | None = None
    value_usd: Decimal | None = None
    fee_usd: Decimal | None = None
    gas_usd: Decimal | None = None
    status: Literal["executed", "rejected"] | None = None
    reason: str | None = None


class Costs(StrictModel):
    total_fees_usd: Decimal | None = None
    total_gas_usd: Decimal | None = None
    total_funding_usd: Decimal | None = None
    total_yield_usd: Decimal | None = None


class ResultMetadata(StrictModel):
    policy: SimulationPolicy
    interval: str | None = None
    start: datetime | None = None
    end: datetime | None = None
    effective_start: datetime | None = None
    effective_end: datetime | None = None
    data_coverage: list[dict[str, Any]] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class BacktestResult(StrictModel):
    schema_version: str = "catalyst.backtest.result.v1"
    summary: Summary
    equity_curve: list[EquityPoint] = Field(default_factory=list)
    drawdown_curve: list[DrawdownPoint] = Field(default_factory=list)
    trades: list[Trade] = Field(default_factory=list)
    final_portfolio: Portfolio | None = None
    costs: Costs | None = None
    metadata: ResultMetadata


__all__ = [
    "BacktestResult",
    "Summary",
    "EquityPoint",
    "DrawdownPoint",
    "Trade",
    "Costs",
    "ResultMetadata",
]

File: crates/contracts/src/result.rs
//! Backtest result contract (backtest-result.schema.json).

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::policy::SimulationPolicy;
use crate::trace::Portfolio;
use crate::Decimal;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Summary {
    pub starting_value_usd: Decimal,
    pub final_value_usd: Decimal,
    pub pnl_usd: Decimal,
    pub return_pct: Decimal,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_drawdown_pct: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub trade_count: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rejected_count: Option<u64>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EquityPoint {
    pub ts: String,
    pub equity_usd: Decimal,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DrawdownPoint {
    pub ts: String,
    pub drawdown_pct: Decimal,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Trade {
    pub ts: String,
    pub node_id: String,
    pub kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub venue: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub symbol: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub side: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub price: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub amount: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub value_usd: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fee_usd: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub gas_usd: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct Costs {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub total_fees_usd: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub total_gas_usd: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub total_funding_usd: Option<Decimal>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub total_yield_usd: Option<Decimal>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ResultMetadata {
    pub policy: SimulationPolicy,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub interval: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub start: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub end: Option<String>,
    /// First actual tick of the run (#167); may be later than the requested `start`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub effective_start: Option<String>,
    /// Last actual tick of the run (#167); may be earlier than the requested `end`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub effective_end: Option<String>,
    #[serde(default)]
    pub data_coverage: Vec<Value>,
    #[serde(default)]
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BacktestResult {
    #[serde(default = "default_result_schema_version")]
    pub schema_version: String,
    pub summary: Summary,
    #[serde(default)]
    pub equity_curve: Vec<EquityPoint>,
    #[serde(default)]
    pub drawdown_curve: Vec<DrawdownPoint>,
    #[serde(default)]
    pub trades: Vec<Trade>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub final_portfolio: Option<Portfolio>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub costs: Option<Costs>,
    pub metadata: ResultMetadata,
}

fn default_result_schema_version() -> String {
    "catalyst.backtest.result.v1".to_string()
}
